[REVIEW] Fix `fillna` & `dropna` to also consider `np.nan` as a missing value

galipremsagar opened the pull request:
Fixes: #7007 

This PR introduces changes to handle the filling of `np.nan` values in `fillna` code by converting `nan` to `null`. This fix surfaced an issue with `can_cast_safely` where when trying to convert a float column with `nan`'s to `int` column is being allowed - This is incorrect and thus added a check to return False if there is atleast 1 `nan` value in the float column.

`nan` is not being handled in `dropna` aswell but is being handled in `isna`, thus introduced changes to `nan` in `dropna` too.

<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

jrhemstad commented on python/cudf/cudf/core/column/column.py:
```diff
@@ -1246,8 +1250,10 @@ def corr(self, other):
     def nans_to_nulls(self):
-        if self.dtype.kind == "f":
-            col = self.fillna(np.nan)
+        if self.dtype.kind == "f" and libcudf.unary.is_nan(self).any():
+            col = libcudf.replace.replace_nulls(
+                self, cudf.Scalar(self.dtype.type(np.nan))
+            )
             newmask = libcudf.transform.nans_to_nulls(col)
```
Maybe I'm misunderstand, but this looks really inefficient. It looks like it first checks if NaNs exist, if they do, replaces nulls with NaN, and then replaces NaNs with nulls?

galipremsagar replied:
Right, after looking back at the history of the code this is now a redundant operation after `nans_to_nulls` got introduced. Removed this part of the code. Thanks for the catch @jrhemstad 🙏

galipremsagar replied:
> That seems to be what is happening. Are we able to do this with a single pass instead? For instance, skip the nan count, skip the replace, and just create a new mask for elements which are valid numeric?

I tried generating the mask in a single go by trying `is_valid`, but `np.nan` is also considered as a valid value by that function which is why had to resort to checking and then extracting the mask.


> Also, is `f` the only type we need to check? What about integers and dates/times?

`nan` is only allowed in our float column, so we don't need that check in rest of the numeric types.

shwina commented on python/cudf/cudf/core/column/column.py:
```diff
@@ -290,8 +290,12 @@ def _concat(cls, objs, dtype=None):
 
-    def dropna(self):
-        dropped_col = self.as_frame().dropna()._as_column()
+    def dropna(self, drop_nan=True):
```
Should we default to `drop_nan=False`? I don't think it's very common to want to drop both nulls and nans at the same time.

galipremsagar replied:
Defaulted this to `False`

cwharris commented:
Once this is fixed how does one replace null values without effecting NaN values?

shwina commented on python/cudf/cudf/core/frame.py:
```diff
@@ -1473,7 +1488,10 @@ def _drop_na_columns(self, how="any", subset=None, thresh=None):
 
         for col in self._data.names:
-            if (len(df[col]) - df[col].null_count) < thresh:
+            if ((len(df[col]) - df[col].null_count) < thresh) or (
+                isinstance(df._data[col], cudf.core.column.NumericalColumn)
+                and ((len(df[col]) - df._data[col].nan_count) < thresh)
+            ):
```
Let's encapsulate what it means for this condition to be true, as it's not obvious from looking at the code:


```python
something = ((len(df[col]) - df[col].null_count) < thresh) or (
                isinstance(df._data[col], cudf.core.column.NumericalColumn)
                and ((len(df[col]) - df._data[col].nan_count) < thresh)
            )

if something:
    ...
```

galipremsagar replied:
Sure, addressed this

cwharris commented:
Can you adjust the title to indicate what fixes are being applied/how fillna/dropna are being improved? That will become the commit message and changelog entry.

galipremsagar commented:
> Once this is fixed how does one replace null values without effecting NaN values?

There is currently no way to achieve this via public API's(`dropna`/`fillna`). We might have to introduce a parameter `drop_nan` similar to the way I have introduced it in the internal APIs.



> Can you adjust the title to indicate what fixes are being applied/how fillna/dropna are being improved? That will become the commit message and changelog entry.

Updated the title. The description seems to be clear, but let me know if I need to change something there to make things clear.

mike-wendt commented:
rerun tests

kkraus14 commented on python/cudf/cudf/core/column/column.py:
```diff
@@ -1248,9 +1254,8 @@ def corr(self, other):
-        if self.dtype.kind == "f":
-            col = self.fillna(np.nan)
-            newmask = libcudf.transform.nans_to_nulls(col)
+        if self.dtype.kind == "f" and libcudf.unary.is_nan(self).any():
```
Isn't this `is_nan.any()` check more expensive than just doing `nans_to_nulls`? `is_nan.any()` would cause unnecessary synchronization.

galipremsagar replied:
Had this chat privately but posting here for visibility, there seems to be huge performance hit when we use `is_nan.any()`, so removed it: 

```python

In [1]: import cudf

In [2]: import numpy as np

In [3]: s = cudf.Series([1, 2, np.nan, None, np.nan, 324, 123, 45] * 10_00_000_00, nan_as_null=False)


# With is_nan.any() check enabled

In [5]: %timeit s.nans_to_nulls()
84.4 ms ± 5.55 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [6]: %timeit s.nans_to_nulls()
85.3 ms ± 4.44 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)




# Without is_nan.any() check


In [4]: %timeit s.nans_to_nulls()
43.6 ms ± 4.72 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [5]: %timeit s.nans_to_nulls()
44.9 ms ± 5.1 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

kkraus14 commented on python/cudf/cudf/core/column/numerical.py:
```diff
@@ -105,6 +105,14 @@ def binary_operator(self, binop, rhs, reflect=False):
+    @property
+    def nan_count(self):
+        if self.dtype.kind == "f":
+            s = cudf.Series(libcudf.unary.is_nan(self))
+            return len(s[s])
```
May be better off doing a sum instead of a boolean masking here.

galipremsagar replied:
We now wouldn't need this function, so removed it.

kkraus14 commented on python/cudf/cudf/core/column/numerical.py:
```diff
@@ -583,6 +595,9 @@ def can_cast_safely(self, to_dtype):
 
         # want to cast float to int:
         elif self.dtype.kind == "f" and to_dtype.kind in {"i", "u"}:
+            if libcudf.unary.is_nan(self).any():
+                return False
```
Instead of this change could we use something like `_drop_na_rows(drop_nan=False)` at line 606?

galipremsagar replied:
Looks like with dropping rows is more expensive than doing an `is_nan.any()` check:
```python
In [1]: import cudf
In [2]: import numpy as np

In [3]: def s1(col):
   ...:     return cudf._lib.unary.is_nan(col).any()
   ...: 

In [4]: def s2(col):
   ...:     return col.dropna(drop_nan=False)
   ...: 

In [5]: s = cudf.Series([1, 2, np.nan, None, np.nan, 324, 123, 45] * 10_00_000_00, nan_as_nul
   ...: l=False)

In [6]: %timeit s1(s._column)
43.8 ms ± 4.74 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [7]: %timeit s2(s._column)
219 ms ± 2.66 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

But an easy solution to this perf issue is to introduce `fill_nan` parameter in `NumericalColumn.fillna`, which will let us control filling of `nan` values. Hence since we are already anyway performing  a `self.fillna`, passing `fill_nan` as `False` will make it more efficient and help us achieve `nan` validation needed here.

kkraus14 commented on python/cudf/cudf/core/frame.py:
```diff
@@ -1501,7 +1516,13 @@ def _drop_na_columns(self, how="any", subset=None, thresh=None):
 
         for col in self._data.names:
-            if (len(df[col]) - df[col].null_count) < thresh:
+            no_threshold_valid_count = (
+                (len(df[col]) - df[col].null_count) < thresh
+            ) or (
+                isinstance(df._data[col], cudf.core.column.NumericalColumn)
+                and ((len(df[col]) - df._data[col].nan_count) < thresh)
+            )
```
May be better off doing a `nan --> null` here and then just checking the thresh once to avoid unnecessary synchronization.

galipremsagar replied:
Done 👍

kkraus14 commented:
@gpucibot merge
